ParseTree

RomanLevashev opened the pull request.

RomanLevashev committed:
ParseTree

YuriUfimtsev commented on ParseTree/ParseTree/Program.cs:
```diff
@@ -0,0 +1,14 @@
+    return;
+}
+
+var sr = new StreamReader(filePath);
```
Названия переменных лучше не сокращать

YuriUfimtsev commented on ParseTree/ParseTree/Operators.cs:
```diff
@@ -0,0 +1,34 @@
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
```
Используйте filse-scoped неймспейсы.
И using-директивы обычно до объявления неймспейса размещают

YuriUfimtsev commented on ParseTree/ParseTree/Nodes/OperationNode.cs:
```diff
@@ -0,0 +1,65 @@
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Text;
+    using System.Threading.Tasks;
```
Аналогично здесь и в других местах

YuriUfimtsev commented on ParseTree/ParseTreeTests/ParseTreeTests.cs:
```diff
@@ -0,0 +1,107 @@
+﻿namespace ParseTreeTests
+{
+    using ParseTree;
```
И здесь

YuriUfimtsev commented on ParseTree/ParseTree/DataHandler.cs:
```diff
@@ -0,0 +1,150 @@
+    /// <summary>
+    /// Provides functionality for parsing and evaluating expressions using parse trees.
+    /// </summary>
+    public class DataHandler
```
Можно static сделать

RomanLevashev committed:
Improvements from comment.
Migrate to .NET 9

RomanLevashev committed:
- Delete StyleCop warnings
- Apply minor changes to comply with style guide
- Rename tests to more descriptive and clear names

yurii-litvinov commented on ParseTree/ParseTree/Nodes/NumberNode.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <returns>
+    /// The integer value stored in this node.
+    /// </returns>
+    public override int Evaluate()
+    {
+        return this.Value;
+    }
```
```suggestion
    public override int Evaluate() => this.Value;
```

yurii-litvinov commented on ParseTree/ParseTree/Nodes/NumberNode.cs:
```diff
@@ -0,0 +1,35 @@
+    /// <summary>
+    /// Prints the numeric value to standard output.
+    /// </summary>
+    public override void Print()
+    {
+        Console.Write(this.Value);
+    }
```
```suggestion
    public override void Print()
        => Console.Write(this.Value);
```

yurii-litvinov commented on ParseTree/ParseTree/Nodes/OperationNode.cs:
```diff
@@ -0,0 +1,66 @@
+    /// </summary>
+    public Operators Operator { get; } = operatorSymbol;
+
+    private static Dictionary<Operators, Func<int, int, int>> OperationsDict { get; } = new()
+    {
+        [Operators.Add] = (a, b) => a + b,
+        [Operators.Subtract] = (a, b) => a - b,
+        [Operators.Multiply] = (a, b) => a * b,
+        [Operators.Divide] = (a, b) => b == 0 ? throw new DivideByZeroException() : a / b,
+    };
```
Это вручную реализованная таблица виртуальных методов. Идеологически более правильно было бы сделать четыре класса-наследника OperationNode, по-своему переопределяющих Evaluate.

yurii-litvinov commented on ParseTree/ParseTree/DataHandler.cs:
```diff
@@ -0,0 +1,150 @@
+            enumerator.MoveNext();
+            return (new NumberNode(value), enumerator);
+        }
+        else if (token == '('.ToString())
```
```suggestion
        else if (token == "(")
```
Странная запись просто. Кажется, мы и так знаем, чему равно '('.ToString()

yurii-litvinov commented on ParseTree/ParseTreeTests/ParseTreeTests.cs:
```diff
@@ -0,0 +1,111 @@
+    /// Tests handling of expressions with missing whitespace between tokens.
+    /// </summary>
+    [TestMethod]
+    public void ThrowFormatExceptionWhenMissingWhitespaceBetweenTokens()
+    {
+        Assert.ThrowsException<FormatException>(() => DataHandler.Parse("(+(* 23) (/ 10 2))"));
+    }
```
```suggestion
    public void ThrowFormatExceptionWhenMissingWhitespaceBetweenTokens()
        => Assert.ThrowsException<FormatException>(() => DataHandler.Parse("(+(* 23) (/ 10 2))"));
```

yurii-litvinov reviewed (approved):
В целом ок. Зачтена.
